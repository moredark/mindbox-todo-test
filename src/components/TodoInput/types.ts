export interface TodoInputProps {
  onAddTodo: (text: string) => void;
}