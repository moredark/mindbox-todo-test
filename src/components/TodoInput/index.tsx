import React, { useState, useCallback, useMemo } from 'react';
import { Box, TextInput, Button } from 'grommet';
import { Add } from 'grommet-icons';
import { TodoInputProps } from './types';

export const TodoInput: React.FC<TodoInputProps> = ({ onAddTodo }) => {
  const [text, setText] = useState('');

  const isValidText = (value: string): boolean => {
    return value.trim().length > 0;
  };

  const handleSubmit = useCallback((e: React.FormEvent) => {
    e.preventDefault();
    if (isValidText(text)) {
      onAddTodo(text.trim());
      setText('');
    }
  }, [text, onAddTodo]);

  const handleKeyDown = useCallback((e: React.KeyboardEvent) => {
    if (e.key === 'Enter' && isValidText(text)) {
      onAddTodo(text.trim());
      setText('');
    }
  }, [text, onAddTodo]);

  const isValid = useMemo(() => isValidText(text), [text]);

  return (
    <Box
      as="form"
      direction="row"
      align="center"
      gap="small"
      margin={{ vertical: 'medium' }}
      onSubmit={handleSubmit}
      data-testid="todo-form"
    >
      <TextInput
        placeholder="What needs to be done?"
        value={text}
        onChange={(e) => setText(e.target.value)}
        onKeyDown={handleKeyDown}
        data-testid="todo-input"
        color="text"
      />
      <Button
        icon={<Add color="light-1" />}
        type="submit"
        disabled={!isValid}
        primary
        data-testid="todo-submit"
      />
    </Box>
  );
};
