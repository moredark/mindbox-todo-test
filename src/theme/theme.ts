import { ThemeType } from 'grommet';
import { deepMerge } from 'grommet/utils';
import { dark } from 'grommet/themes';

export const accentColor = '#7D4CDB';

export const customTheme: ThemeType = deepMerge(dark, {
  global: {
    colors: {
      background: '#121212',
      'background-back': '#1E1E1E',
      'background-front': '#2D2D2D',
      'background-contrast': '#3D3D3D',
      text: '#EEEEEE',
      'text-strong': '#FFFFFF',
      'dark-6': '#999999',
      'dark-5': '#BBBBBB',
      'accent-1': accentColor,
      brand: accentColor
    },
    focus: {
      border: {
        color: accentColor
      }
    }
  },
  button: {
    border: {
      radius: '4px'
    },
    primary: {
      color: accentColor
    }
  },
  checkBox: {
    hover: {
      border: {
        color: accentColor
      }
    },
    color: accentColor
  }
});
