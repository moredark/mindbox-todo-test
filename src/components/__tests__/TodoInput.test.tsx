import { render, screen, fireEvent } from '@testing-library/react';
import { Grommet } from 'grommet';
import { TodoInput } from '../TodoInput';
import { customTheme } from '../../theme/theme';

describe('TodoInput', () => {
  const defaultProps = {
    onAddTodo: jest.fn()
  };

  const renderComponent = (props = defaultProps) => {
    return render(
      <Grommet theme={customTheme}>
        <TodoInput {...props} />
      </Grommet>
    );
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should update input value when typing', () => {
    renderComponent();
    const input = screen.getByTestId('todo-input');

    fireEvent.change(input, { target: { value: 'New todo' } });

    expect(input).toHaveValue('New todo');
  });

  it('should clear input after submitting valid text', () => {
    renderComponent();
    const input = screen.getByTestId('todo-input');
    const form = screen.getByTestId('todo-form');

    fireEvent.change(input, { target: { value: 'New todo' } });
    fireEvent.submit(form);

    expect(input).toHaveValue('');
    expect(defaultProps.onAddTodo).toHaveBeenCalledWith('New todo');
  });

  it('should not call onAddTodo when submitting empty text', () => {
    renderComponent();
    const form = screen.getByTestId('todo-form');

    fireEvent.submit(form);

    expect(defaultProps.onAddTodo).not.toHaveBeenCalled();
  });

  it('should not call onAddTodo when submitting whitespace', () => {
    renderComponent();
    const input = screen.getByTestId('todo-input');
    const form = screen.getByTestId('todo-form');

    fireEvent.change(input, { target: { value: '   ' } });
    fireEvent.submit(form);

    expect(defaultProps.onAddTodo).not.toHaveBeenCalled();
  });

  it('should trim text before adding todo', () => {
    renderComponent();
    const input = screen.getByTestId('todo-input');
    const form = screen.getByTestId('todo-form');

    fireEvent.change(input, { target: { value: '  New todo  ' } });
    fireEvent.submit(form);

    expect(defaultProps.onAddTodo).toHaveBeenCalledWith('New todo');
  });

  it('should add todo on Enter key press with valid text', () => {
    renderComponent();
    const input = screen.getByTestId('todo-input');

    fireEvent.change(input, { target: { value: 'New todo' } });
    fireEvent.keyDown(input, { key: 'Enter' });

    expect(input).toHaveValue('');
    expect(defaultProps.onAddTodo).toHaveBeenCalledWith('New todo');
  });

  it('should not add todo on Enter key press with invalid text', () => {
    renderComponent();
    const input = screen.getByTestId('todo-input');

    fireEvent.change(input, { target: { value: '  ' } });
    fireEvent.keyDown(input, { key: 'Enter' });

    expect(defaultProps.onAddTodo).not.toHaveBeenCalled();
  });

  it('should disable submit button when text is empty', () => {
    renderComponent();
    const submitButton = screen.getByTestId('todo-submit');

    expect(submitButton).toBeDisabled();
  });

  it('should enable submit button when text is valid', () => {
    renderComponent();
    const input = screen.getByTestId('todo-input');
    const submitButton = screen.getByTestId('todo-submit');

    fireEvent.change(input, { target: { value: 'New todo' } });

    expect(submitButton).not.toBeDisabled();
  });
});
