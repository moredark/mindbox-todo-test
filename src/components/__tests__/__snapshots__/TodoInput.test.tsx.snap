// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoInput should match snapshot 1`] = `
<div>
  <div
    class="StyledGrommet-sc-19lkkz7-0 lioFhE"
  >
    <form
      class="StyledBox-sc-13pk1d4-0 eqbLyG"
      data-testid="todo-form"
    >
      <div
        class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 gcUiwQ"
      >
        <input
          autocomplete="off"
          class="StyledTextInput-sc-1x30a0s-0 eHTvhq"
          color="text"
          data-testid="todo-input"
          placeholder="What needs to be done?"
          value=""
        />
      </div>
      <div
        class="StyledBox__StyledBoxGap-sc-13pk1d4-1 hDDuS"
      />
      <button
        class="StyledButton-sc-323bzc-0 ceKFSL"
        data-testid="todo-submit"
        disabled=""
        type="submit"
      >
        <svg
          aria-label="Add"
          class="StyledIcon-sc-ofa7kd-0 eyecLz"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 22V2M2 12h20"
            fill="none"
            stroke="#000"
            stroke-width="2"
          />
        </svg>
      </button>
    </form>
  </div>
</div>
`;
